src/test/drop_select_packets.rs

```diff
@@ -0,0 +1,156 @@
+
+    let res = core.run(future::lazy(move || {
+        let (mut plug_client, mut plug_server) = IpPlug::new_pair();
+
+        let mut i = 0;
+        let filter = move |packet| {
+            if let IpPacket::V4(ref ipv4_packet) = packet {
+                if let Ipv4Payload::Udp(udp_packet) = ipv4_packet.payload() {
+                    let data = BytesMut::from(udp_packet.payload());
+                    if let Ok(p) = Packet::parse(data) {
+                        let prev_i = i;
+                        i += 1;
+                        if prev_i == n {
+                            trace!("dropping packet: {:?}", p);
+                            return None;
+                        }
+                    }
+                }
+            }
+            Some(packet)
+        };
+
+        if sender {
+            plug_server = plug_server.filter_map(filter).into_ip_plug(&network_handle);
+        } else {
+            plug_client = plug_client.filter_map(filter).into_ip_plug(&network_handle);
+        }
+
+        let ip_client = Ipv4Addr::random_global();
+        let spawn_client = netsim::device::MachineBuilder::new()
+            .add_ip_iface(
+                netsim::iface::IpIfaceBuilder::new()
+                    .ipv4_addr(ip_client, 0)
+                    .ipv4_route(Ipv4Route::new(Ipv4Range::global(), None)),
+                plug_client,
+            )
+            .spawn(&network_handle, move || {
+                let mut core = unwrap!(Core::new());
+                let handle = core.handle();
+                let (socket_client, _) = unwrap!(UtpSocket::bind(&addr!("0.0.0.0:0"), &handle));
+                let their_addr = unwrap!(addr_rx.recv());
+                let res = core.run(future::lazy(move || {
+                    socket_client
+                        .connect(&their_addr)
+                        .map_err(|e| {
+                            panic!("connect error: {}", e);
+                        })
+                        .and_then(|stream_client| {
+                            tokio_io::io::write_all(stream_client, b"ping")
+                                .map_err(|e| panic!("write error: {}", e))
+                                .and_then(|(stream_client, _bytes)| {
+                                    tokio_io::io::read_to_end(stream_client, Vec::new())
+                                        .map_err(|e| panic!("read error: {}", e))
+                                        .and_then(|(stream_client, bytes)| {
+                                            assert_eq!(&bytes[..], b"pong");
+                                            tokio_io::io::shutdown(stream_client)
+                                                .map_err(|e| panic!("error shuting down: {}", e))
+                                                .and_then(|stream_client| {
+                                                    stream_client.finalize().infallible()
+                                                })
```

Comment:
Is this not exactly we want to avoid with futures style code ? How is it different from say a callback-hell ?

Similarly in other places (e.g., below)